relations/src/preimage.rs

```diff
@@ -0,0 +1,141 @@
+use poseidon::r1cs;
+
+use crate::{
+    environment::FpVar,
+    relation::state::{FullInput, NoInput, OnlyPublicInput, State, WithPublicInput},
+    CircuitField, GetPublicInput,
+};
+
+// Poseidon paper suggests using domain separation for this, concretely encoding the use case in the capacity element (which is fine as it is 256 bits large and has a lot of bits to fill)
+static DOMAIN_SEPARATOR: Lazy<Fr> = Lazy::new(|| Fr::from(2137));
+
+/// Preimage relation : H(preimage)=hash
+/// where:
+/// - hash : public input
+/// - preimage : private witness
+#[derive(Copy, Clone, Eq, PartialEq, Hash, Debug)]
+pub struct PreimageRelation<S: State> {
+    // private witness
+    pub preimage: Option<CircuitField>,
+    // public input
+    pub hash: Option<CircuitField>,
+    _phantom: PhantomData<S>,
+}
+
+impl PreimageRelation<NoInput> {
+    pub fn without_input() -> Self {
+        Self {
+            hash: None,
+            preimage: None,
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl PreimageRelation<OnlyPublicInput> {
+    pub fn with_public_input(hash: CircuitField) -> Self {
+        Self {
+            preimage: None,
+            hash: Some(hash),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl PreimageRelation<FullInput> {
+    pub fn with_full_input(preimage: CircuitField, hash: CircuitField) -> Self {
+        Self {
+            preimage: Some(preimage),
+            hash: Some(hash),
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<S: State> ConstraintSynthesizer<CircuitField> for PreimageRelation<S> {
+    fn generate_constraints(
+        self,
+        cs: ConstraintSystemRef<CircuitField>,
+    ) -> Result<(), SynthesisError> {
+        let preimage = FpVar::new_witness(ns!(cs, "preimage"), || {
+            self.preimage.ok_or(AssignmentMissing)
+        })?;
+        let hash = FpVar::new_input(ns!(cs, "hash"), || self.hash.ok_or(AssignmentMissing))?;
+
+        let domain_separator = FpVar::new_constant(cs.clone(), *DOMAIN_SEPARATOR)?;
+        let hash_result = r1cs::one_to_one_hash(cs, &domain_separator, preimage)?;
+
+        hash.enforce_equal(&hash_result)?;
+
+        Ok(())
+    }
+}
+
+impl<S: WithPublicInput> GetPublicInput<CircuitField> for PreimageRelation<S> {
+    fn public_input(&self) -> Vec<CircuitField> {
+        vec![self
+            .hash
+            .expect("Circuit should have public input assigned")]
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use ark_bls12_381::Bls12_381;
+    use ark_crypto_primitives::SNARK;
+    use ark_groth16::Groth16;
+    use ark_relations::r1cs::{ConstraintSynthesizer, ConstraintSystem};
+    use poseidon::hash;
+
+    use super::{PreimageRelation, DOMAIN_SEPARATOR};
+    use crate::{CircuitField, GetPublicInput};
+
+    #[test]
+    fn preimage_constraints_correctness() {
+        let preimage = CircuitField::from(17u64);
+        let preimage_hash = hash::one_to_one_hash(&DOMAIN_SEPARATOR, preimage);
```

Comment:
For a mathematician, it seems enormously disturbing to not name this var image :P

Reply:
![iguess](https://user-images.githubusercontent.com/468572/211834036-b10a2273-3d54-42a3-a0a9-53538d7486d1.png)